materialize-snowflake/pipe.go

```diff
@@ -0,0 +1,267 @@
+package main
+
+import (
+	"bytes"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"github.com/golang-jwt/jwt/v5"
+	"github.com/google/uuid"
+	log "github.com/sirupsen/logrus"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"text/template"
+	"time"
+)
+
+type PipeClient struct {
+	cfg             *config
+	base            string
+	token           string
+	httpClient      http.Client
+	insertFilesTpl  *template.Template
+	insertReportTpl *template.Template
+	account         string
+}
+
+func publicKeyFingerprint(publicKey *rsa.PublicKey) (string, error) {
+	der, err := x509.MarshalPKIXPublicKey(publicKey)
+	if err != nil {
+		return "", fmt.Errorf("marshalling public key: %w", err)
+	}
+	var hash = sha256.Sum256(der)
+	return fmt.Sprintf("SHA256:%s", base64.StdEncoding.EncodeToString(hash[:])), nil
+}
+
+func NewPipeClient(cfg *config, tenant string) (*PipeClient, error) {
+	httpClient := http.Client{}
+
+	var key, err = cfg.Credentials.privateKey()
+	if err != nil {
+		return nil, err
+	}
+
+	var account = strings.ToUpper(cfg.Account)
+	var user = strings.ToUpper(cfg.Credentials.User)
+	var qualifiedUser = fmt.Sprintf("%s.%s", account, user)
+
+	fingerprint, err := publicKeyFingerprint(key.Public().(*rsa.PublicKey))
+	if err != nil {
+		return nil, err
+	}
+
+	var claims = &jwt.RegisteredClaims{
+		IssuedAt: jwt.NewNumericDate(time.Now()),
+		// TODO: automatically refresh the JWT token
+		// JWT tokens for Snowflake can live up to an hour
+		ExpiresAt: jwt.NewNumericDate(time.Now().Add(59 * time.Minute)),
+		Issuer:    fmt.Sprintf("%s.%s", qualifiedUser, fingerprint),
+		Subject:   qualifiedUser,
```

Comment:
Pondering this some more: I'd really like to get rid of the `account` configuration input completely, and just require the `host` config input. It seems like you actually do need the `account` here, and it would be nice if the [`CURRENT_ACCOUNT_NAME`](https://docs.snowflake.com/en/sql-reference/functions/current_account_name) (or maybe `CURRENT_ACCOUNT`?) function could be used for getting that with the `host`-based connection, rather than requiring it as configuration.